### DISCUSSION TITLE
Use GPU with kubernetes yaml support in podman.

### DISCUSSION THREAD

**Discussion by andrew-kennedy** [containers/podman#12587] [Q&A]
I've successfully used gpu support in the docker-style container creation shell shell incantations within podman, for applications like plex and jellyfin running in containers, as I've installed the nvidia-container-runtime and with the specification of the proper runtime/gpu flags, everything works.

I'm trying to move my podman pods configuration into kubernetes yaml, for ease of eventual transition to k8s. Is there a good way to use GPUs when configuring containers to be run with `podman play kube`?

**Comment by rhatdan** [comment]
Would love to have you write an article on what you have done.
@umohnani8 @haircommander @mrunalp @saschagrunert Could you answer the question on kube?

**Comment by saschagrunert** [comment]
We could use the Kubernetes resource indicators to setup the correct runtime and its flags:
```yaml
      resources:
        limits:
          nvidia.com/gpu: 1 # requesting 1 GPU
```
This should work in the same way for AMD and Intel graphics.

**Comment by fpoirotte** [comment]
See also https://github.com/containers/podman/issues/17833 (and https://github.com/containers/podman/discussions/14934).

As pointed out by @saschagrunert, support for GPUs (or other types of devices) in Kubernetes v1.26+ uses the [Device Plugins Framework](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/): you install a vendor-supplied plugin on your cluster, then request the device as a resource in your YAML file through a vendor-specific keyword (similar to how annotations work).

Implementing the framework in podman may prove difficult if support for every existing vendor is required.
E.g. in my case, I would like to use the [Kubernetes Generic device plugin](https://github.com/squat/generic-device-plugin/) which uses completely different keywords (by default, these are named `squat.ai/video`, `squat.ai/serial`, `squat.ai/fuse`, `squat.ai/audio`, `squat.ai/capture`, but [the names can be customized and new keywords added as well](https://github.com/squat/generic-device-plugin/blob/main/manifests/generic-device-plugin.yaml)).
